torch/testing/_internal/common_utils.py

```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                                                            enable_zero_sized=enable_zero_sized,
+                                                            enable_batched_variable_nse=enable_batched_variable_nse,
+                                                            output_tensor=False):
+                if layout is torch.strided:
+                    size = kwargs.pop('size', args[0].shape)  # to ensure that a zero-sized tensor has the desired shape
+                    assert len(args) == 1
+                    yield args[0].reshape(size)
+                elif layout is torch.sparse_coo:
+                    yield torch.sparse_coo_tensor(*args, **kwargs)
+                elif is_compressed_sparse_layout:
+                    kwargs.update(layout=layout)
+                    yield torch.sparse_compressed_tensor(*args, **kwargs)
+                else:
+                    assert 0  # unreachable
+            return
+
+        def get_shape(pattern):
+            if isinstance(pattern, list):
+                if not pattern:
+                    return (0,)
+                return (len(pattern), *get_shape(pattern[0]))
+            return ()
+
+        def get_blockpattern(pattern, blocksize):
+            basesize = get_shape(pattern)
+            assert basesize[0] % blocksize[0] == 0, (basesize, blocksize)
+            assert basesize[1] % blocksize[1] == 0, (basesize, blocksize)
+            blockpattern = [[0 for _ in range(basesize[1] // blocksize[1])] for _ in range(basesize[0] // blocksize[0])]
+            for i, row in enumerate(pattern):
+                for j, maskin in enumerate(row):
+                    bi = i // blocksize[0]
+                    bj = j // blocksize[1]
+                    blockno = 1 + bi * (basesize[1] // blocksize[1]) + bj
+                    blockpattern[bi][bj] = blockno if maskin else blockpattern[bi][bj]
+            return blockpattern
+
+        def transpose(lst):
+            return list(map(list, zip(*lst)))
+
+        def get_sparse_data(pattern):
+            indices = []
+            values = []
+            count = 0
+            size = get_shape(pattern)
+            batchsize = size[:-2]
+            basesize = size[-2:]
+            crow_indices = [0 for _ in range(basesize[0] + 1)]
+            ccol_indices = [0 for _ in range(basesize[1] + 1)]
+            strided_values = [[0 for _ in range(basesize[1])] for _ in range(basesize[0])]
+            for i, row in enumerate(pattern):
+                for j, maskin in enumerate(row):
+                    if maskin:
+                        count += 1
+                        indices.append([i, j])
+                        values.append(count)
+                        strided_values[i][j] = count
+                        for i1 in range(i, basesize[0]):
+                            crow_indices[i1 + 1] += 1
+                        for j1 in range(j, basesize[1]):
+                            ccol_indices[j1 + 1] += 1
```

Comment:
nit: We can probably simplify this with tensors + masking + advanced indexing.

Reply:
generate_simple_inputs now uses tensors